solidity/test/vault/TBTCVault.Redemption.test.ts

```diff
@@ -0,0 +1,580 @@
+    await bank.setBalance(account2.address, initialBankBalance)
+    await bank
+      .connect(account1)
+      .approveBalance(tbtcVault.address, initialBankBalance)
+    await bank
+      .connect(account2)
+      .approveBalance(tbtcVault.address, initialBankBalance)
+
+    await bridge.setWallet(walletPubKeyHash, {
+      ecdsaWalletID: ethers.constants.HashZero,
+      mainUtxoHash: ethers.constants.HashZero,
+      pendingRedemptionsValue: 0,
+      createdAt: await lastBlockTime(),
+      movingFundsRequestedAt: 0,
+      closingStartedAt: 0,
+      pendingMovedFundsSweepRequestsCount: 0,
+      state: walletState.Live,
+      movingFundsTargetWalletsCommitmentHash: ethers.constants.HashZero,
+    })
+    await bridge.setWalletMainUtxo(walletPubKeyHash, mainUtxo)
+  })
+
+  describe("redeem", () => {
+    const requestRedemption = async (
+      redeemer: SignerWithAddress,
+      redeemerOutputScript: string,
+      amount: number
+    ): Promise<ContractTransaction> => {
+      const data = defaultAbiCoder.encode(
+        ["address", "bytes20", "bytes32", "uint32", "uint64", "bytes"],
+        [
+          redeemer.address,
+          walletPubKeyHash,
+          mainUtxo.txHash,
+          mainUtxo.txOutputIndex,
+          mainUtxo.txOutputValue,
+          redeemerOutputScript,
+        ]
+      )
+
+      return tbtcVault.connect(redeemer).redeem(amount, data)
+    }
+
+    context("when the redeemer has no TBTC", () => {
+      const amount = to1e18(1)
+      before(async () => {
+        await createSnapshot()
+
+        await tbtc.connect(account1).approve(tbtcVault.address, amount)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should revert", async () => {
+        await expect(
+          tbtcVault.connect(account1).redeem(to1e18(1), [])
+        ).to.be.revertedWith("Burn amount exceeds balance")
+      })
+    })
+
+    context("when the redeemer has not enough TBTC", () => {
+      const mintedAmount = to1e18(1)
+      const redeemedAmount = mintedAmount.add(1)
+
+      before(async () => {
+        await createSnapshot()
+
+        await tbtcVault.connect(account1).mint(mintedAmount)
+        await tbtc.connect(account1).approve(tbtcVault.address, redeemedAmount)
+      })
+
+      after(async () => {
+        await restoreSnapshot()
+      })
+
+      it("should revert", async () => {
+        await expect(
+          tbtcVault.connect(account1).redeem(redeemedAmount, [])
+        ).to.be.revertedWith("Burn amount exceeds balance")
+      })
+    })
+
+    context("when there is a single redeemer", () => {
+      const redeemerOutputScriptP2WPKH =
+        "0x160014f4eedc8f40d4b8e30771f792b065ebec0abaddef"
+      const redeemerOutputScriptP2WSH =
+        "0x220020ef0b4d985752aa5ef6243e4c6f6bebc2a007e7d671ef27d4b1d0db8dcc93bc1c"
+      const redeemerOutputScriptP2PKH =
+        "0x1976a914f4eedc8f40d4b8e30771f792b065ebec0abaddef88ac"
```

Comment:
Let's also test the `P2SH` case.

Reply:
Makes sense! See 08228f958a885618b33e435dd7253dd872fe3fe2.